### PR TITLE
[TASK-7] initial impl

### PR DESCRIPTION
**Pull request by Nikkov17**
FE: https://d1u63n4q8vz46v.cloudfront.net/

Import url: https://th8mexnih6.execute-api.eu-west-1.amazonaws.com/dev/import
Header title: Authorization
Header value: Tmlra292MTc6VEVTVF9QQVNTV09SRA==

To check functionality you should add 'authorization_token' field to the localStorage of the deployed FE with Authorization header value above.

Basic done:
1 - authorization-service is added to the repo, has correct basicAuthorizer lambda and correct serverless.yaml file
3 - import-service serverless.yaml file has authorizer configuration for the importProductsFile lambda. Request to the importProductsFile lambda should work only with correct authorization_token being decoded and checked by basicAuthorizer lambda. Response should be in 403 HTTP status if access is denied for this user (invalid authorization_token) and in 401 HTTP status if Authorization header is not provided.
5 - update client application to send Authorization: Basic authorization_token header on import. Client should get authorization_token value from browser localStorage https://developer.mozilla.org/ru/docs/Web/API/Window/localStorage authorization_token = localStorage.getItem('authorization_token')

Estimated score:
5/6

### CODE REVIEW

**Comment by Nikkov17** [comment]
Hi to reviewers!
I have a question about additional task with alerts:
As you can see on the frontend, if authorization_token in localStorage contains wrong token, in response we have just failed status, without any statusCode or error message, so how can we handle this error to get statusCode(401 or 403) and show correct alert?

**Comment by AndreyS55** [comment]
> Hi to reviewers!
> I have a question about additional task with alerts:
> As you can see on the frontend, if authorization_token in localStorage contains wrong token, in response we have just failed status, without any statusCode or error message, so how can we handle this error to get statusCode(401 or 403) and show correct alert?

As I understand the response blocked by CORS. That's why you don't see correct statuses. I don't see `basicAuthorizer` lambda integration into `import-service`. You should adjust `importProductsFile` lambda from `import-service` with authorizer. Also you can add default API Gateway responses in your `import-service` to avoid CORS errors.

File: authorization-service/basicauthorizer.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+module.exports.basicAuthorizer = async (event, context, cb) => {
+  console.log(`Event: ${JSON.stringify(event)}`);
+
+  if (event.type !== "TOKEN") {
+    cb("Unauthorized");
+  }
+
+  try {
+    const authorizationToken = event.authorizationToken;
```
Comment (AndreyS55) [inline authorization-service/basicauthorizer.js:13]: I suggest to add next check
```
if (!authorizationToken) cb('Unauthorized');
```

File: authorization-service/basicauthorizer.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    const encodedCreds = authorizationToken.split(" ")[1];
```
Comment (AndreyS55) [inline authorization-service/basicauthorizer.js:15]: I also suggest to check that `authorizationToken.split(" ")[0] === 'Basic'` and `!authorizationToken.split(" ")[1]` and throw `Unauthorized` if they are not valid.

**Comment by kotmatpockuh** [comment]
I guess base64 encoding should also be done on FE side... Not to put base64 key into localstorage 🤷‍♂️

**Comment by kotmatpockuh** [comment]
Also looks like when the key is invalid - there is no error alert... Please fix it.

**Comment by alexeyMohnatkin** [comment]
no `401` or `403` errors, adding `GatewayResponseAccessDeied` and `GatewayResponseUnauthorized` to the serverless config should resolve it

File: authorization-service/basicauthorizer.js (hunk 1 above)
Comment (Nikkov17) [reply authorization-service/basicauthorizer.js:13]: Good suggestion, thanks! will implement it during next task!

File: authorization-service/basicauthorizer.js (hunk 2 above)
Comment (Nikkov17) [reply authorization-service/basicauthorizer.js:15]: Another one good suggestion, thanks!

**Comment by Nikkov17** [comment]
> I guess base64 encoding should also be done on FE side... Not to put base64 key into localstorage 🤷‍♂️

Hmm, maybe you're right, it would be more convenient

**Comment by Nikkov17** [comment]
> > Hi to reviewers!
> > I have a question about additional task with alerts:
> > As you can see on the frontend, if authorization_token in localStorage contains wrong token, in response we have just failed status, without any statusCode or error message, so how can we handle this error to get statusCode(401 or 403) and show correct alert?
> 
> As I understand the response blocked by CORS. That's why you don't see correct statuses. I don't see `basicAuthorizer` lambda integration into `import-service`. You should adjust `importProductsFile` lambda from `import-service` with authorizer. Also you can add default API Gateway responses in your `import-service` to avoid CORS errors.

Thanks for info about CORS, maybe i didn't check the console and miss CORS errors)
I did integration of authorizer in the aws console, not using serverless file. 
As far as i can see we should add something like this in serverless file.
  events:
    ...
    authorizer: basicAuthorizer
  authorizerFunc:
    handler: handler.basicAuthorizer
but use arn for basicAuthorizer to get it from another service

 i will add it and cors fix in the scope of next task.

**Pull request closed without merging**